test/kafka-upsert-benchmark/mzworkflows.py

```diff
@@ -0,0 +1,165 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+import time
+from typing import List
+
+import pg8000
+
+from materialize import ui
+from materialize.mzcompose import (
+    Kafka,
+    Kgen,
+    Materialized,
+    Testdrive,
+    Workflow,
+    WorkflowArgumentParser,
+    Zookeeper,
+    say,
+)
+from materialize.ui import UIError
+
+mz_persist = Materialized(
+    name="materialized_persist",
+    hostname="materialized",
+    options="--persistent-user-tables --persistent-kafka-upsert-source --disable-persistent-system-tables-test",
+)
+
+mz_default = Materialized()
+
+prerequisites = ["zookeeper", "kafka"]
+
+services = [
+    Zookeeper(),
+    Kafka(),
+    mz_default,
+    mz_persist,
+    # This instance of Testdrive is used to initialize the benchmark, and needs to have a
+    # known seed.
+    Testdrive(no_reset=True, seed=1),
+    Kgen(),
+]
+
+
+def kafka_upsert_benchmark_step(
+    w: Workflow,
+    mz_service: str,
+    restart_mz: bool,
+    records_sent: int,
+    num_records: int,
+    num_keys: int,
+    value_bytes: int,
+    timeout_secs: int,
+):
+    if restart_mz:
```

Comment:
I am not sure the `--restart-mz` is measuring what you expect -- it measures the time it takes for Mz to recover/re-ingest previously-ingested records **and** catch up with the new records just inserted via kgen. Therefore, each measurement is measuring progressively more and more work being done:

With this command:

```
./mzcompose run  kafka-upsert-benchmark --num-steps 1000 --records-per-step 1000000 --num-keys 10000000000 --restart-mz 
```

I get the following measurements:

```
C> query result: 1200000 after 6.007135797291994
C> query result: 2200000 after 14.013093177229166
C> query result: 3200000 after 24.01748914644122
C> query result: 4200000 after 31.023660700768232
C> query result: 5200000 after 43.02803621068597
```

With persistence enabled, it is a similar story.